paimon-core/src/main/java/org/apache/paimon/utils/AsyncRecordReader.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.paimon.utils;
+
+import org.apache.paimon.reader.RecordReader;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.LinkedBlockingQueue;
+
+/** A {@link RecordReader} to use ASYNC_EXECUTOR to read records async. */
+public class AsyncRecordReader<T> implements RecordReader<T> {
+
+    private static final ExecutorService ASYNC_EXECUTOR =
+            Executors.newCachedThreadPool(new ExecutorThreadFactory("paimon-reader-async-thread"));
+
+    private final BlockingQueue<Element> queue;
+    private final Future<Void> future;
+
+    private boolean isEnd = false;
+
+    public AsyncRecordReader(IOExceptionSupplier<RecordReader<T>> supplier) {
+        this.queue = new LinkedBlockingQueue<>();
+        this.future =
+                ASYNC_EXECUTOR.submit(
+                        () -> {
+                            try (RecordReader<T> reader = supplier.get()) {
```

Comment:
Maybe we can throw the exception, and announce the worker thread to stop to avoid worker thread blocked at queue.get() forever.